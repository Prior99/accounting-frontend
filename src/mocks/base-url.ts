declare var global: any;
global.baseUrl = "example.com";
