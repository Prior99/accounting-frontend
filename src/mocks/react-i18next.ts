import * as React from "react";

jest.mock("react-i18next", () => ({
    translate: (namespaces?: string[], options?: any) => <C extends Function>(WrappedComponent: C) => {
        class Translated extends React.Component<{}, undefined> {
            public render() {
                const props = Object.assign({}, this.props, {
                    t: (name: string) => name,
                });
                return React.createElement(WrappedComponent as any, props);
            }
        }
        return Translated;
    },
}));
