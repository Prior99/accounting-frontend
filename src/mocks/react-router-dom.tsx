import * as React from "react";

jest.mock("react-router-dom", () => ({
    Link: ({ to, children }: { to: string, children: JSX.Element }) => (<a href={to}>{children}</a>),
}));
