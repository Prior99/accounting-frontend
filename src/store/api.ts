import { store } from ".";

declare var baseUrl: string;

type HTTPMethod = "DELETE" | "GET" | "PUT" | "POST";

export async function api(url: string, body?: any, method: HTTPMethod = "GET"): Promise<any> {
    const headers = new Headers();
    if (store.login.loggedIn) {
        headers.append("authorization", store.login.authToken);
    }
    headers.append("content-type", "application/json");
    const fullUrl = `${window.location.protocol}//${baseUrl}${url}`;
    try {
        const response = await fetch(fullUrl, {
            method,
            headers,
            body: JSON.stringify(body),
        });
        const jsonResponse = (await response.json());
        const okay = response.status >= 200 && response.status <= 299;
        if (!okay) {
            store.error.apiError(`Fetching from url "${fullUrl}" resulted in error: "${jsonResponse.message}"`);
            return { okay };
        }
        return {
            okay,
            data: jsonResponse.data,
        };
    } catch (err) {
        store.error.apiError(`Failed to fetch from url "${fullUrl}".`, err);
        return { okay: false };
    }
}
