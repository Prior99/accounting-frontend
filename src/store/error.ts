import { observable, computed, action } from "mobx";

export interface DisplayableError {
    dismissed: boolean;
    message: string;
    origin: "api" | "internal";
    error?: Error;
}

export class ErrorStore {
    constructor() {
        this.apiError = this.apiError.bind(this);
        this.internalError = this.internalError.bind(this);
        this.dismissCurrent = this.dismissCurrent.bind(this);
    }

    @observable public errors: DisplayableError[] = [];

    @computed public get current() {
        return this.errors.find(error => !error.dismissed);
    }

    @action
    public dismissCurrent() {
        if (typeof this.current === "undefined") {
            return;
        }
        this.current.dismissed = true;
    }

    @action
    public apiError(message: string, error?: Error) {
        this.errors.push({ dismissed: false, message, origin: "api", error });
    }

    @action
    public internalError(message: string, error?: Error) {
        this.errors.push({ dismissed: false, message, origin: "internal", error });
    }
}
