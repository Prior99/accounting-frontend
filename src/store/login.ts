import { observable, computed, action } from "mobx";
import { api } from "./api";
import { bind } from "bind-decorator";
import { browserHistory } from "browser-history";
import { routeDashboard } from "routing";
import { RequestStatus } from "request-status";

const softwareVersion = 1;
const localStorageIdentifier = "accounting-software-login";

interface LocalStorageLogin {
    storageVersion: number;
    date: string;
    authToken: string;
}

export class LoginStore {
    @observable public authToken: string;
    @observable public status = RequestStatus.PENDING;

    constructor() {
        this.load();
    }

    @computed
    public get loggedIn() {
        return typeof this.authToken !== "undefined";
    }

    @bind
    @action
    public async onLogin(email: string, password: string) {
        this.status = RequestStatus.IN_PROGRESS;
        const { data, okay } = await api("/auth/login", { email, password }, "POST", true);
        if (okay && data.token) {
            this.authToken = data.token;
            this.save();
            browserHistory.replace(routeDashboard());
            this.status = RequestStatus.SUCCESS;
        } else {
            this.status = RequestStatus.FAIL;
        }
    }

    @bind
    public save() {
        const deserialized: LocalStorageLogin = {
            storageVersion: softwareVersion,
            date: new Date().toString(),
            authToken: this.authToken,
        };
        const serialized = JSON.stringify(deserialized);
        localStorage.setItem(localStorageIdentifier, serialized);
    }

    @bind
    public clearStorage() {
        localStorage.removeItem(localStorageIdentifier);
    }

    @bind
    public load() {
        const serialized = localStorage.getItem(localStorageIdentifier);
        if (serialized === null) { // tslint:disable-line
            return;
        }
        let deserialized: LocalStorageLogin;
        try {
            deserialized = JSON.parse(serialized);
        } catch (err) {
            this.clearStorage();
            return;
        }
        if (deserialized.storageVersion !== softwareVersion) {
            this.clearStorage();
            return;
        }
        this.authToken = deserialized.authToken;
        browserHistory.replace(routeDashboard());
    }
}
