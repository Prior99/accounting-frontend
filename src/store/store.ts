import { ApiStore } from "./api";
import { History } from "history";

export class Store {
    protected api: ApiStore;
    protected browserHistory: History;

    constructor(api: ApiStore, browserHistory: History) {
        this.api = api;
        this.browserHistory = browserHistory;
    }
}
