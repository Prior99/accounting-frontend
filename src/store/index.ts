import { LoginStore } from "./login";
import { SignupStore } from "./signup";
import { ErrorStore } from "./error";
export * from "./error";

export interface Store {
    login: LoginStore;
    error: ErrorStore;
    signup: SignupStore;
}

export let store: Store;

export function createStore(): Store {
    return {
        login: new LoginStore(),
        error: new ErrorStore(),
        signup: new SignupStore()
    };
}

export function resetStore() {
    store = createStore();
}

resetStore();

if (typeof window !== "undefined") {
    (window as any).store = store;
    (window as any).createStore = createStore;
    (window as any).resetStore = resetStore;
}
