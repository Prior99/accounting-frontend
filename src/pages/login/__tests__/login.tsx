import "mocks/react-router-dom";
import "mocks/react-i18next";

import * as React from "react";
import { StrippedPageLogin, mapStoreToProps } from "..";
import { mount } from "enzyme";
import toJson from "enzyme-to-json";
import { store, resetStore } from "store";

beforeEach(() => {
    resetStore();
});

test("`PageLogin` is rendered correctly initially", () => {
    const mounted = mount(<StrippedPageLogin failed={false} onLogin={jest.fn()} />);
    expect(toJson(mounted)).toMatchSnapshot();
    expect(mounted.find("button").props()["disabled"]).toBe(true);
});

test("`PageLogin` is rendered with a warning message if the login is failed", () => {
    const mounted = mount(<StrippedPageLogin failed={true}  onLogin={jest.fn()}/>);
    expect(toJson(mounted)).toMatchSnapshot();
});

test("`PageLogin` is rendered correctly with all fields filled out", () => {
    const mounted = mount(<StrippedPageLogin failed={false} onLogin={jest.fn()} />);
    mounted.find("input[placeholder='email']").simulate("change", { target: { value: "test@example.com" }});
    mounted.find("input[placeholder='password']").simulate("change", { target: { value: "abcdefgh" }});
    expect(toJson(mounted)).toMatchSnapshot();
    expect(mounted.find("Button").props()["disabled"]).toBe(false);
});

test("`mapStoreToProps()` returns the expected properties", () => {
    const props = mapStoreToProps(store);
    expect(props).toEqual({
        onLogin: store.login.onLogin,
        failed: false
    });
});

test("`PageLogin` calls `onLogin` if the button is clicked", () => {
    const onLogin = jest.fn();
    const mounted = mount(<StrippedPageLogin failed={false} onLogin={onLogin}/>);
    mounted.find("form").simulate("submit");
    expect(onLogin).toHaveBeenCalled();
});
