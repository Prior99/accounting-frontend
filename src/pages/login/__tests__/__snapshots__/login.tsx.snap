// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`\`PageLogin\` is rendered correctly initially 1`] = `
<Translated
  onLogin={[Function]}
  status="pending"
>
  <StrippedPageLogin
    onLogin={[Function]}
    status="pending"
    t={[Function]}
  >
    <Grid
      centered={true}
      style={
        Object {
          "margin": 0,
        }
      }
      verticalAlign="middle"
    >
      <div
        className="ui centered middle aligned grid"
        style={
          Object {
            "margin": 0,
          }
        }
      >
        <GridColumn
          stretched={true}
        >
          <div
            className="stretched column"
          >
            <h1>
              common:appName
            </h1>
            <Translated
              status="pending"
            >
              <StatusMessage
                status="pending"
                t={[Function]}
              />
            </Translated>
            <Segment
              stacked={true}
            >
              <div
                className="ui stacked segment"
              >
                <Form
                  as="form"
                  onSubmit={[Function]}
                  size="large"
                >
                  <form
                    className="ui large form"
                    onSubmit={[Function]}
                  >
                    <FormField>
                      <div
                        className="field"
                      >
                        <Input
                          error={true}
                          focus={true}
                          icon="user"
                          iconPosition="left"
                          onChange={[Function]}
                          placeholder="email"
                          size="large"
                          type="text"
                          value=""
                        >
                          <div
                            className="ui large error focus left icon input"
                          >
                            <Icon
                              as="i"
                              name="user"
                            >
                              <i
                                aria-hidden="true"
                                className="user icon"
                              />
                            </Icon>
                            <input
                              onChange={[Function]}
                              placeholder="email"
                              type="text"
                              value=""
                            />
                          </div>
                        </Input>
                      </div>
                    </FormField>
                    <FormField>
                      <div
                        className="field"
                      >
                        <Input
                          error={true}
                          focus={true}
                          icon="lock"
                          iconPosition="left"
                          onChange={[Function]}
                          placeholder="password"
                          size="large"
                          type="password"
                          value=""
                        >
                          <div
                            className="ui large error focus left icon input"
                          >
                            <Icon
                              as="i"
                              name="lock"
                            >
                              <i
                                aria-hidden="true"
                                className="lock icon"
                              />
                            </Icon>
                            <input
                              onChange={[Function]}
                              placeholder="password"
                              type="password"
                              value=""
                            />
                          </div>
                        </Input>
                      </div>
                    </FormField>
                    <Button
                      as="button"
                      color="olive"
                      disabled={true}
                      fluid={true}
                      type="submit"
                    >
                      <button
                        className="ui olive fluid disabled button"
                        disabled={true}
                        onClick={[Function]}
                        tabIndex={-1}
                        type="submit"
                      >
                        login
                      </button>
                    </Button>
                  </form>
                </Form>
              </div>
            </Segment>
            <Segment
              tertiary={true}
            >
              <div
                className="ui tertiary segment"
              >
                noAccount
                 
                <Link
                  to="/signup"
                >
                  <a
                    href="/signup"
                  >
                    signup
                  </a>
                </Link>
              </div>
            </Segment>
          </div>
        </GridColumn>
      </div>
    </Grid>
  </StrippedPageLogin>
</Translated>
`;

exports[`\`PageLogin\` is rendered correctly with all fields filled out 1`] = `
<Translated
  onLogin={[Function]}
  status="pending"
>
  <StrippedPageLogin
    onLogin={[Function]}
    status="pending"
    t={[Function]}
  >
    <Grid
      centered={true}
      style={
        Object {
          "margin": 0,
        }
      }
      verticalAlign="middle"
    >
      <div
        className="ui centered middle aligned grid"
        style={
          Object {
            "margin": 0,
          }
        }
      >
        <GridColumn
          stretched={true}
        >
          <div
            className="stretched column"
          >
            <h1>
              common:appName
            </h1>
            <Translated
              status="pending"
            >
              <StatusMessage
                status="pending"
                t={[Function]}
              />
            </Translated>
            <Segment
              stacked={true}
            >
              <div
                className="ui stacked segment"
              >
                <Form
                  as="form"
                  onSubmit={[Function]}
                  size="large"
                >
                  <form
                    className="ui large form"
                    onSubmit={[Function]}
                  >
                    <FormField>
                      <div
                        className="field"
                      >
                        <Input
                          error={false}
                          focus={true}
                          icon="user"
                          iconPosition="left"
                          onChange={[Function]}
                          placeholder="email"
                          size="large"
                          type="text"
                          value="test@example.com"
                        >
                          <div
                            className="ui large focus left icon input"
                          >
                            <Icon
                              as="i"
                              name="user"
                            >
                              <i
                                aria-hidden="true"
                                className="user icon"
                              />
                            </Icon>
                            <input
                              onChange={[Function]}
                              placeholder="email"
                              type="text"
                              value="test@example.com"
                            />
                          </div>
                        </Input>
                      </div>
                    </FormField>
                    <FormField>
                      <div
                        className="field"
                      >
                        <Input
                          error={false}
                          focus={true}
                          icon="lock"
                          iconPosition="left"
                          onChange={[Function]}
                          placeholder="password"
                          size="large"
                          type="password"
                          value="abcdefgh"
                        >
                          <div
                            className="ui large focus left icon input"
                          >
                            <Icon
                              as="i"
                              name="lock"
                            >
                              <i
                                aria-hidden="true"
                                className="lock icon"
                              />
                            </Icon>
                            <input
                              onChange={[Function]}
                              placeholder="password"
                              type="password"
                              value="abcdefgh"
                            />
                          </div>
                        </Input>
                      </div>
                    </FormField>
                    <Button
                      as="button"
                      color="olive"
                      disabled={false}
                      fluid={true}
                      type="submit"
                    >
                      <button
                        className="ui olive fluid button"
                        onClick={[Function]}
                        type="submit"
                      >
                        login
                      </button>
                    </Button>
                  </form>
                </Form>
              </div>
            </Segment>
            <Segment
              tertiary={true}
            >
              <div
                className="ui tertiary segment"
              >
                noAccount
                 
                <Link
                  to="/signup"
                >
                  <a
                    href="/signup"
                  >
                    signup
                  </a>
                </Link>
              </div>
            </Segment>
          </div>
        </GridColumn>
      </div>
    </Grid>
  </StrippedPageLogin>
</Translated>
`;

exports[`\`PageLogin\` is rendered with a warning message if the login is failed 1`] = `
<Translated
  onLogin={[Function]}
  status="fail"
>
  <StrippedPageLogin
    onLogin={[Function]}
    status="fail"
    t={[Function]}
  >
    <Grid
      centered={true}
      style={
        Object {
          "margin": 0,
        }
      }
      verticalAlign="middle"
    >
      <div
        className="ui centered middle aligned grid"
        style={
          Object {
            "margin": 0,
          }
        }
      >
        <GridColumn
          stretched={true}
        >
          <div
            className="stretched column"
          >
            <h1>
              common:appName
            </h1>
            <Translated
              status="fail"
            >
              <StatusMessage
                status="fail"
                t={[Function]}
              >
                <Message
                  content="loginFailed.content"
                  header="loginFailed.headline"
                  icon="warning sign"
                  warning={true}
                >
                  <div
                    className="ui icon warning message"
                  >
                    <Icon
                      as="i"
                      name="warning sign"
                    >
                      <i
                        aria-hidden="true"
                        className="warning sign icon"
                      />
                    </Icon>
                    <MessageContent>
                      <div
                        className="content"
                      >
                        <MessageHeader
                          content="loginFailed.headline"
                        >
                          <div
                            className="header"
                          >
                            loginFailed.headline
                          </div>
                        </MessageHeader>
                        <p>
                          loginFailed.content
                        </p>
                      </div>
                    </MessageContent>
                  </div>
                </Message>
              </StatusMessage>
            </Translated>
            <Segment
              stacked={true}
            >
              <div
                className="ui stacked segment"
              >
                <Form
                  as="form"
                  onSubmit={[Function]}
                  size="large"
                >
                  <form
                    className="ui large form"
                    onSubmit={[Function]}
                  >
                    <FormField>
                      <div
                        className="field"
                      >
                        <Input
                          error={true}
                          focus={true}
                          icon="user"
                          iconPosition="left"
                          onChange={[Function]}
                          placeholder="email"
                          size="large"
                          type="text"
                          value=""
                        >
                          <div
                            className="ui large error focus left icon input"
                          >
                            <Icon
                              as="i"
                              name="user"
                            >
                              <i
                                aria-hidden="true"
                                className="user icon"
                              />
                            </Icon>
                            <input
                              onChange={[Function]}
                              placeholder="email"
                              type="text"
                              value=""
                            />
                          </div>
                        </Input>
                      </div>
                    </FormField>
                    <FormField>
                      <div
                        className="field"
                      >
                        <Input
                          error={true}
                          focus={true}
                          icon="lock"
                          iconPosition="left"
                          onChange={[Function]}
                          placeholder="password"
                          size="large"
                          type="password"
                          value=""
                        >
                          <div
                            className="ui large error focus left icon input"
                          >
                            <Icon
                              as="i"
                              name="lock"
                            >
                              <i
                                aria-hidden="true"
                                className="lock icon"
                              />
                            </Icon>
                            <input
                              onChange={[Function]}
                              placeholder="password"
                              type="password"
                              value=""
                            />
                          </div>
                        </Input>
                      </div>
                    </FormField>
                    <Button
                      as="button"
                      color="olive"
                      disabled={true}
                      fluid={true}
                      type="submit"
                    >
                      <button
                        className="ui olive fluid disabled button"
                        disabled={true}
                        onClick={[Function]}
                        tabIndex={-1}
                        type="submit"
                      >
                        login
                      </button>
                    </Button>
                  </form>
                </Form>
              </div>
            </Segment>
            <Segment
              tertiary={true}
            >
              <div
                className="ui tertiary segment"
              >
                noAccount
                 
                <Link
                  to="/signup"
                >
                  <a
                    href="/signup"
                  >
                    signup
                  </a>
                </Link>
              </div>
            </Segment>
          </div>
        </GridColumn>
      </div>
    </Grid>
  </StrippedPageLogin>
</Translated>
`;
